src/transport/raw/TCP.cpp

```diff
@@ -290,101 +268,82 @@ CHIP_ERROR TCPBase::SendAfterConnect(const PeerAddress & addr, System::PacketBuf
     return err;
 }
 
-CHIP_ERROR TCPBase::ProcessSingleMessageFromBufferHead(const PeerAddress & peerAddress, const System::PacketBufferHandle & buffer,
-                                                       uint16_t messageSize)
-{
-    CHIP_ERROR err     = CHIP_NO_ERROR;
-    uint8_t * oldStart = buffer->Start();
-    uint16_t oldLength = buffer->DataLength();
-
-    buffer->SetDataLength(messageSize);
-
-    PacketHeader header;
-    err = header.DecodeAndConsume(buffer);
-    SuccessOrExit(err);
-
-    // message receive handler will attempt to free the buffer, however as the buffer may
-    // contain additional data, we retain it to prevent actual free
-    HandleMessageReceived(header, peerAddress, buffer.Retain());
-
-exit:
-    buffer->SetStart(oldStart);
-    buffer->SetDataLength(oldLength);
-
-    return err;
-}
-
 CHIP_ERROR TCPBase::ProcessReceivedBuffer(Inet::TCPEndPoint * endPoint, const PeerAddress & peerAddress,
                                           System::PacketBufferHandle buffer)
 {
-    CHIP_ERROR err = CHIP_NO_ERROR;
+    CHIP_ERROR err  = CHIP_NO_ERROR;
+    size_t consumed = 0;
 
     while (!buffer.IsNull())
     {
-        // when a buffer is empty, it can be released back to the app
-        if (buffer->DataLength() == 0)
+        uint8_t messageSizeBuf[kPacketSizeBytes];
+        if (buffer->Read(messageSizeBuf, sizeof(messageSizeBuf)) != CHIP_NO_ERROR)
         {
-            buffer.FreeHead();
-            continue;
+            break;
         }
-
-        uint8_t * messageData = nullptr;
-        uint16_t messageSize  = 0;
-        if (ContainsCompleteMessage(buffer, &messageData, &messageSize))
+        uint16_t messageSize = LittleEndian::Get16(messageSizeBuf);
+        if (messageSize + kPacketSizeBytes <= buffer->TotalLength())
```

Comment:
I keep worrying about overflow when I see this sort of thing.  On a system with a 16-bit size_t this would be a problem... though we may not have such systems.  And on other systems this relies on the precise integer promotion rules to not fail, I think.

How would you feel about either adding a static_assert that `size_t` is large enough that this should never overflow or rewriting as:
```
if (messageSize <= buffer->TotalLength() - kPacketSizeBytes)
```
?  This should be safe, because if `TotalLength()` were less than `kMessageSizeBytes` the `Read` would have failed, right?

Reply:
We need to error-out if `messageSize >= (UINT16_MAX - sizeof(uint16))` as it would be impossible to hold that in a TotalLength of the chain.